usrsctplib/netinet/sctp_callout.c

```diff
@@ -114,154 +174,197 @@ sctp_os_timer_start(sctp_os_timer_t *c, int to_ticks, void (*ftn) (void *),
+
+void
+sctp_os_timer_cancel(sctp_os_timer_t* c)
 {
-	int wakeup_cookie;
+	SCTP_TIMERQ_LOCK();
 
+	switch (c->c_state)
+	{
+	case SCTP_CALLOUT_SCHEDULED:
+	{
+		sctp_binary_heap_remove(&SCTP_BASE_INFO(timers_queue), &c->heap_node);
+		c->c_state = SCTP_CALLOUT_COMPLETED;
+		break;
+	}
+	case SCTP_CALLOUT_RUNNING:
+	{
+		c->c_state = SCTP_CALLOUT_CANCEL_REQUESTED;
+		break;
+	}
+	case SCTP_CALLOUT_CANCEL_REQUESTED:
+	case SCTP_CALLOUT_COMPLETED:
+	case SCTP_CALLOUT_NEW:
+	{
+		// nothing to do
+		break;
+	}
+	default:
+	{
+		KASSERT(0, ("Unknown callout state"));
+		break;
+	}
+	}
+
+	SCTP_TIMERQ_UNLOCK();
+}
+
+
+void
+sctp_os_timer_wait_completion(sctp_os_timer_t* c)
+{
 	SCTP_TIMERQ_LOCK();
-	/*
-	 * Don't attempt to delete a callout that's not on the queue.
-	 */
-	if (!(c->c_flags & SCTP_CALLOUT_PENDING)) {
-		c->c_flags &= ~SCTP_CALLOUT_ACTIVE;
-		if (sctp_os_timer_current != c) {
-			SCTP_TIMERQ_UNLOCK();
-			return (0);
-		} else {
-			/*
-			 * Deleting the callout from the currently running
-			 * callout from the same thread, so just return
-			 */
-			userland_thread_id_t tid;
-			sctp_userspace_thread_id(&tid);
-			if (sctp_userspace_thread_equal(tid,
-						sctp_os_timer_current_tid)) {
-				SCTP_TIMERQ_UNLOCK();
-				return (0);
-			}
 
-			/* need to wait until the callout is finished */
-			sctp_os_timer_waiting = 1;
-			wakeup_cookie = ++sctp_os_timer_wait_ctr;
-			SCTP_TIMERQ_UNLOCK();
-			SCTP_TIMERWAIT_LOCK();
-			/*
-			 * wait only if sctp_handle_tick didn't do a wakeup
-			 * in between the lock dance
-			 */
-			if (wakeup_cookie - sctp_os_timer_done_ctr > 0) {
-#if defined (__Userspace_os_Windows)
-				SleepConditionVariableCS(&sctp_os_timer_wait_cond,
-							 &sctp_os_timerwait_mtx,
-							 INFINITE);
-#else
-				pthread_cond_wait(&sctp_os_timer_wait_cond,
-						  &sctp_os_timerwait_mtx);
+	userland_thread_id_t current_tid;
+	sctp_userspace_thread_id(&current_tid);
+
+	int retry_condition_wait = 1;
+	do
+	{
+		switch (c->c_state)
+		{
+		case SCTP_CALLOUT_RUNNING:
+		case SCTP_CALLOUT_CANCEL_REQUESTED:
+		case SCTP_CALLOUT_SCHEDULED:
+		{
+			if (current_tid == c->c_executor_id)
+			{
+				// callout tried to wait for completion of itself
+				retry_condition_wait = 0;
+			}
+			else
+			{
+#if defined (__Userspace__)
+				sctp_userland_cond_wait(&c->c_completion, &SCTP_BASE_VAR(timer_mtx));
 #endif
 			}
-			SCTP_TIMERWAIT_UNLOCK();
+			break;
 		}
-		return (0);
-	}
-	c->c_flags &= ~(SCTP_CALLOUT_ACTIVE | SCTP_CALLOUT_PENDING);
-	if (c == sctp_os_timer_next) {
-		sctp_os_timer_next = TAILQ_NEXT(c, tqe);
-	}
-	TAILQ_REMOVE(&SCTP_BASE_INFO(callqueue), c, tqe);
+		case SCTP_CALLOUT_COMPLETED:
+		case SCTP_CALLOUT_NEW:
+		{
+			retry_condition_wait = 0;
+			break;
+		}
+		default:
+		{
+			KASSERT(0, ("Unknown callout state"));
+			break;
+		}
+		}
+	} while (retry_condition_wait);
+
+	SCTP_TIMERQ_UNLOCK();
+}
+
+
+int
+sctp_os_timer_stop(sctp_os_timer_t *c)
+{
+	SCTP_TIMERQ_LOCK();
```

Comment:
Is recursive locking enabled for this, on all platforms? From a quick look it seemed like it's not.

Reply:
Thank you very much! I've fixed that by extracting several helper functions which relies assumes that mutex is held by holder.